sycl/unittests/assert/assert.cpp

```diff
@@ -0,0 +1,385 @@
+  Mock.redefine<PiApiKind::piProgramBuild>(redefinedProgramBuild);
+  Mock.redefine<PiApiKind::piKernelCreate>(redefinedKernelCreate);
+  Mock.redefine<PiApiKind::piKernelSetExecInfo>(redefinedKernelSetExecInfo);
+  Mock.redefine<PiApiKind::piKernelGetInfo>(redefinedKernelGetInfo);
+  Mock.redefine<PiApiKind::piKernelGetGroupInfo>(redefinedKernelGetGroupInfo);
+  Mock.redefine<PiApiKind::piEnqueueKernelLaunch>(redefinedEnqueueKernelLaunch);
+  Mock.redefine<PiApiKind::piMemBufferCreate>(redefinedMemBufferCreate);
+  Mock.redefine<PiApiKind::piMemRelease>(redefinedMemRelease);
+  Mock.redefine<PiApiKind::piProgramRetain>(redefinedProgramRetain);
+  Mock.redefine<PiApiKind::piKernelRetain>(redefinedKernelRetain);
+  Mock.redefine<PiApiKind::piKernelRelease>(redefinedKernelRelease);
+  Mock.redefine<PiApiKind::piKernelSetArg>(redefinedKernelSetArg);
+  Mock.redefine<PiApiKind::piEnqueueMemBufferMap>(redefinedEnqueueMemBufferMap);
+  Mock.redefine<PiApiKind::piEventsWait>(redefinedEventsWait);
+  Mock.redefine<PiApiKind::piextKernelSetArgMemObj>(
+      redefinedExtKernelSetArgMemObj);
+}
+
+TEST(Assert, Test) {
+  sycl::platform Plt{sycl::default_selector()};
+  if (Plt.is_host()) {
+    std::cerr << "Test is not supported on host, skipping\n";
+    return; // test is not supported on host.
+  }
+
+  if (Plt.get_backend() == sycl::backend::cuda) {
+    std::cerr << "Test is not supported on CUDA platform, skipping\n";
+    return;
+  }
+
+  sycl::unittest::PiMock Mock{Plt};
+
+  setupMock(Mock);
+
+  std::atomic<bool> ErrorCaptured{false};
+
+  const sycl::device Dev = Plt.get_devices()[0];
+  sycl::queue Queue{Dev, [&](sycl::exception_list EL) {
+                      for (auto &EPtr : EL)
+                        try {
+                          std::rethrow_exception(EPtr);
+                        } catch (sycl::exception &E) {
+                          if (E.get_cl_code() == PI_ERROR_UNKNOWN)
+                            ErrorCaptured = true;
+                        }
+                    }};
+
+  const sycl::context Ctx = Queue.get_context();
+
+  sycl::kernel_bundle KernelBundle =
+      sycl::get_kernel_bundle<sycl::bundle_state::input>(Ctx, {Dev});
+  auto ExecBundle = sycl::build(KernelBundle);
+  Queue.submit([&](sycl::handler &H) {
+    H.use_kernel_bundle(ExecBundle);
+    H.single_task<TestKernel>([] {});
+  });
+
+  while (!StartedWait)
+    ;
+
+  ContinueWait = true;
+
+  // Can't return from redefinedEventsWait and report atomically. Hence, here
+  // is this wait. Single second wait should be more than enough.
+  {
+    using namespace std::chrono_literals;
+    std::this_thread::sleep_for(1000ms);
+  }
+
+  Queue.throw_asynchronous();
+
+  while (!ErrorCaptured)
+    ;
+
+  // Host-task didn't finish as we returned PI_ERROR_UNKNOWN
+  EXPECT_EQ(ErrorCaptured, true);
+  // Two kernels to be enqueued: the test kernel and assert info copier
+  EXPECT_EQ(KernelLaunchCounter, 2);
```

Comment:
Could you please check the host task prints assert message? Ok, if done in a separate PR.

Reply:
Done